File: crates/octobot_commons/src/tree/base_tree.rs
use std::collections::HashMap;
use std::fmt;

#[derive(Debug)]
pub struct NodeExistsError;

impl fmt::Display for NodeExistsError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.write_str("NodeExistsError: node does not exist at the given path")
    }
}

impl std::error::Error for NodeExistsError {}

/// Node element of a BaseTree. Generic over value type `T`.
///
/// In the pure Rust crate, `T` is typically `NodeValue`.
/// In the PyO3 bridge, `T` is `Py<PyAny>` for arbitrary Python objects.
pub struct BaseTreeNode<T> {
    pub node_value: Option<T>,
    pub node_value_time: f64,
    pub node_type: Option<T>,
    pub node_description: Option<T>,
    pub node_metadata: HashMap<String, T>,
    pub children: HashMap<String, BaseTreeNode<T>>,
}

impl<T> BaseTreeNode<T> {
    pub fn new(node_value: Option<T>, node_type: Option<T>) -> Self {
        Self {
            node_value,
            node_value_time: 0.0,
            node_type,
            node_description: None,
            node_metadata: HashMap::new(),
            children: HashMap::new(),
        }
    }

    pub fn set_child(&mut self, key: String, child: BaseTreeNode<T>) {
        self.children.insert(key, child);
    }

    pub fn pop_child(&mut self, key: &str) -> Option<BaseTreeNode<T>> {
        self.children.remove(key)
    }
}

impl<T> Default for BaseTreeNode<T> {
    fn default() -> Self {
        Self::new(None, None)
    }
}

/// Tree based on BaseTreeNode.
pub struct BaseTree<T> {
    pub root: BaseTreeNode<T>,
}

impl<T> BaseTree<T> {
    pub fn new() -> Self {
        Self {
            root: BaseTreeNode::default(),
        }
    }

    /// Set attributes on an existing node.
    pub fn set_node(
        node: &mut BaseTreeNode<T>,
        value: Option<T>,
        node_type: Option<T>,
        timestamp: f64,
    ) {
        Self::apply_values(node, value, node_type, timestamp, None, None);
    }

    /// Set node at path, creating intermediate nodes as needed.
    pub fn set_node_at_path(
        &mut self,
        value: Option<T>,
        node_type: Option<T>,
        path: &[String],
        timestamp: f64,
        description: Option<T>,
        metadata: Option<HashMap<String, T>>,
    ) {
        let node = self.get_or_create_node(path);
        Self::apply_values(node, value, node_type, timestamp, description, metadata);
    }

    /// Get immutable node at path. Returns error if path does not exist.
    pub fn get_node<'a>(
        &'a self,
        path: &[String],
        starting_node: Option<&'a BaseTreeNode<T>>,
    ) -> Result<&'a BaseTreeNode<T>, NodeExistsError> {
        let start = starting_node.unwrap_or(&self.root);
        Self::traverse(path, start)
    }

    /// Get mutable node at path, creating it if it does not exist.
    pub fn get_or_create_node(&mut self, path: &[String]) -> &mut BaseTreeNode<T> {
        let mut current = &mut self.root;
        for key in path {
            if !current.children.contains_key(key) {
                current
                    .children
                    .insert(key.clone(), BaseTreeNode::default());
            }
            current = current.children.get_mut(key).unwrap();
        }
        current
    }

    /// Get mutable node at path relative to a starting node index path.
    /// Use when you need a mutable reference from a known sub-path.
    pub fn get_or_create_node_from(
        &mut self,
        base_path: &[String],
        sub_path: &[String],
    ) -> &mut BaseTreeNode<T> {
        let full: Vec<String> = base_path
            .iter()
            .chain(sub_path.iter())
            .cloned()
            .collect();
        self.get_or_create_node(&full)
    }

    /// Delete node at path. Returns the deleted node or error.
    pub fn delete_node(
        &mut self,
        path: &[String],
    ) -> Result<BaseTreeNode<T>, NodeExistsError> {
        if path.is_empty() {
            return Err(NodeExistsError);
        }
        let parent_path = &path[..path.len() - 1];
        let key = &path[path.len() - 1];

        let parent = Self::traverse_mut(parent_path, &mut self.root)?;
        parent.pop_child(key).ok_or(NodeExistsError)
    }

    /// Get children keys of node at path.
    pub fn get_children_keys(
        &self,
        path: &[String],
    ) -> Result<Vec<String>, NodeExistsError> {
        let node = self.get_node(path, None)?;
        Ok(node.children.keys().cloned().collect())
    }

    /// Depth-first traversal yielding (node_ref, path) pairs.
    pub fn get_nested_children_with_path<'a>(
        &'a self,
        path: &[String],
        select_leaves_only: bool,
    ) -> Result<Vec<(&'a BaseTreeNode<T>, Vec<String>)>, NodeExistsError> {
        #![allow(clippy::type_complexity)]
        let mut results = Vec::new();
        self.collect_nested(path, select_leaves_only, &mut results)?;
        Ok(results)
    }

    /// Clear the tree.
    pub fn clear(&mut self) {
        self.root = BaseTreeNode::default();
    }

    // --- Private helpers ---

    fn traverse<'a>(
        path: &[String],
        start: &'a BaseTreeNode<T>,
    ) -> Result<&'a BaseTreeNode<T>, NodeExistsError> {
        let mut current = start;
        for key in path {
            current = current.children.get(key).ok_or(NodeExistsError)?;
        }
        Ok(current)
    }

    fn traverse_mut<'a>(
        path: &[String],
        start: &'a mut BaseTreeNode<T>,
    ) -> Result<&'a mut BaseTreeNode<T>, NodeExistsError> {
        let mut current = start;
        for key in path {
            current = current.children.get_mut(key).ok_or(NodeExistsError)?;
        }
        Ok(current)
    }

    fn apply_values(
        node: &mut BaseTreeNode<T>,
        value: Option<T>,
        node_type: Option<T>,
        timestamp: f64,
        description: Option<T>,
        metadata: Option<HashMap<String, T>>,
    ) {
        if value.is_some() {
            node.node_value = value;
        }
        if node_type.is_some() {
            node.node_type = node_type;
        }
        node.node_value_time = timestamp;
        if description.is_some() {
            node.node_description = description;
        }
        if let Some(meta) = metadata {
            node.node_metadata = meta;
        }
    }

    fn collect_nested<'a>(
        &'a self,
        parent_path: &[String],
        select_leaves_only: bool,
        results: &mut Vec<(&'a BaseTreeNode<T>, Vec<String>)>,
    ) -> Result<(), NodeExistsError> {
        let node = self.get_node(parent_path, None)?;
        let children_keys: Vec<String> = node.children.keys().cloned().collect();

        if children_keys.is_empty() || !select_leaves_only {
            results.push((node, parent_path.to_vec()));
        }

        for key in children_keys {
            let mut child_path = parent_path.to_vec();
            child_path.push(key);
            self.collect_nested(&child_path, select_leaves_only, results)?;
        }
        Ok(())
    }
}

impl<T> Default for BaseTree<T> {
    fn default() -> Self {
        Self::new()
    }
}

File: packages/commons/octobot_commons/tree/base_tree.py
#  Drakkar-Software OctoBot-Commons
#  Copyright (c) Drakkar-Software, All rights reserved.
#
#  This library is free software; you can redistribute it and/or
#  modify it under the terms of the GNU Lesser General Public
#  License as published by the Free Software Foundation; either
#  version 3.0 of the License, or (at your option) any later version.
#
#  This library is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
#  Lesser General Public License for more details.
#
#  You should have received a copy of the GNU Lesser General Public
#  License along with this library.


class BaseTreeNode:
    """
    Node element of a BaseTree
    """

    __slots__ = [
        "node_value",
        "node_value_time",
        "node_type",
        "node_description",
        "node_metadata",
        "children",
    ]

    def __init__(self, node_value, node_type, **_):
        self.node_value = node_value
        self.node_value_time = None
        self.node_type = node_type
        self.node_description = None
        self.node_metadata = {}
        self.children = {}

    def set_child(self, key, child):
        """
        Set a child at the given key
        """
        self.children[key] = child

    def pop_child(self, key, default):
        """
        Pop the child the given key
        """
        return self.children.pop(key, default)


class NodeExistsError(Exception):
    """
    Node doesn't exist error
    """


class BaseTree:
    """
    Tree based on BaseTreeNode
    """

    TREE_NODE_CLASS = BaseTreeNode
    __slots__ = ["root"]

    def __init__(self):
        """
        Init the root node
        """
        self.root = self.TREE_NODE_CLASS(None, None)

    def set_node(self, value, node_type, node, timestamp=0):
        """
        Set the node attributes
        Can raise an exception if the node doesn't exists
        :param value: the node 'node_value' attribute to set
        :param node_type: the node 'node_type' attribute to set
        :param node: the node to update
        :param timestamp: the value modification timestamp.
        """
        self._set_node(node, value, node_type, timestamp=timestamp)

    # pylint: disable=too-many-arguments
    def set_node_at_path(
        self, value, node_type, path, timestamp=0, description=None, metadata=None
    ):
        """
        Set the node attributes
        Creates the node if it doesn't exists
        :param value: the node 'node_value' attribute to set
        :param node_type: the node 'node_type' attribute to set
        :param path: the node path (as a list of string)
        :param timestamp: the value modification timestamp.
        :param description: the node 'node_description' attribute to set
        :param metadata: the node 'node_metadata' attribute to set
        For example:
        - If you created a first node with the path ["my-parent-node"]
        - You can create a child node of my-parent-node by using ["my-parent-node", "my-new-child-node"] as `path`
        :return: void
        """
        self._set_node(
            self.get_or_create_node(path),
            value,
            node_type,
            timestamp=timestamp,
            description=description,
            metadata=metadata,
        )

    def get_node(self, path, starting_node=None):
        """
        Get the node at the specified path
        :param path: the node path (as a list of string).
        For example:
        - If you created a first node with the path ["my-parent-node"]
        - You can create a child node of my-parent-node by using ["my-parent-node", "my-new-child-node"] as `path`
        :param starting_node: the node to start the relative path
        :return: the node instance or raise a NodeExistsError if the node doesn't exists
        """
        try:
            return self._get_node(path, starting_node=starting_node)
        except KeyError:
            raise NodeExistsError

    def clear(self):
        """
        Clears the whole tree
        """
        self.root = self.TREE_NODE_CLASS(None, None)

    def delete_node(self, path, starting_node=None):
        """
        Delete the node at the specified path
        :param path: the node path (as a list of string).
        For example:
        - If you created a first node with the path ["my-parent-node"]
        - You can create a child node of my-parent-node by using ["my-parent-node", "my-new-child-node"] as `path`
        :param starting_node: the node to start the relative path
        :return: the deleted node or raise a NodeExistsError if the node doesn't exists
        """
        try:
            deleted_node = self._delete_node(path, starting_node=starting_node)
            if deleted_node is None:
                raise NodeExistsError
            return deleted_node
        except KeyError:
            raise NodeExistsError

    def get_or_create_node(self, path, starting_node=None, **kwargs):
        """
        Get the node at the specified path
        Creates the node if it doesn't exists
        :param path: the node path (as a list of string).
        For example:
        - If you created a first node with the path ["my-parent-node"]
        - You can create a child node of my-parent-node by using ["my-parent-node", "my-new-child-node"] as `path`
        :param starting_node: the node to start the relative path
        :return: the node instance
        """
        try:
            return self._get_node(path, starting_node=starting_node)
        except KeyError:
            return self._create_node_path(path, starting_node=starting_node, **kwargs)

    def get_nested_children_with_path(self, path=None, select_leaves_only=True):
        """
        Returns a generator iterating over the nodes children, including nested children. Children are yielded
        together with their node path using a depth-first search (the most nested children are returned first)
        :param path: the path (as a list of string) to the node
        :param select_leaves_only: when True (default), only nodes that don't have children are returned
        :return: a generator of (node, path) tuples
        """
        path = path or []
        return self._get_nested_children_with_path(path, select_leaves_only)

    def _get_nested_children_with_path(self, parent_path, select_leaves_only):
        children_keys = self.get_children_keys(parent_path)
        node = self.get_node(parent_path)
        if not children_keys or not select_leaves_only:
            yield node, parent_path
        for key in children_keys:
            path = list(parent_path)
            path.append(key)
            yield from self._get_nested_children_with_path(path, select_leaves_only)

    def get_children_keys(self, path):
        """
        Return the node's children keys
        Can raise a KeyError if the path does not exists
        :param path: the path (as a list of string) to the node
        :return: children keys as a list
        """
        return list(self.get_node(path).children)

    def _get_node(self, path, starting_node=None):
        """
        Return the node corresponding to the path
        Can raise a KeyError if the path does not exists
        :param path: the path (as a list of string) to the node
        :param starting_node: the node to start the path, root if None
        :return: BaseTreeNode at path
        """
        current_node = self.root if starting_node is None else starting_node
        for key in path:
            current_node = current_node.children[key]
        return current_node

    def _delete_node(self, path, starting_node=None):
        """
        Return the node corresponding to the path
        Can raise a KeyError if the path does not exists
        :param path: the path (as a list of string) to the node
        :param starting_node: the node to start the path, root if None
        :return: BaseTreeNode at path
        """
        current_node = self.root if starting_node is None else starting_node
        for key in path[:-1]:
            current_node = current_node.children[key]
        return current_node.pop_child(path[-1], None)

    def _create_node_path(self, path, starting_node=None, **kwargs):
        """
        Expensive method that creates the path to the selected node
        :param path: path (as a list of string) to the selected node
        :param starting_node: the node to start the path, root if None
        :return: the created node path
        """
        current_node = self.root if starting_node is None else starting_node
        for key in path:
            try:
                current_node = current_node.children[key]
            except KeyError:
                # create a new node as the current node child
                # us it as the new node
                current_node = self.child_factory(current_node, key, **kwargs)

        return current_node

    def child_factory(self, node, key, **kwargs):
        """
        Create a new child an associate it to the given key
        """
        node.set_child(key, self.TREE_NODE_CLASS(None, None, **kwargs))
        return node.children[key]

    # pylint: disable=too-many-arguments
    def _set_node(
        self,
        node,
        value=None,
        node_type=None,
        timestamp=0,
        description=None,
        metadata=None,
    ):
        """
        Sets the node attributes
        :param node: the node instance to update
        :param value: the node instance 'node_value' attribute to set (is ignored if None)
        :param node_type: the node instance 'node_type' attribute to set (is ignored if None)
        :param timestamp: the value modification timestamp.
        :param description: the node instance 'node_description' attribute to set (is ignored if None)
        :param metadata: the node instance 'node_metadata' attribute to set (is ignored if None)
        """
        if value is not None:
            node.node_value = value

        if node_type is not None:
            node.node_type = node_type

        # set the node value modification timestamp
        node.node_value_time = timestamp

        if description is not None:
            node.node_description = description

        if metadata is not None:
            node.node_metadata = metadata
